packages/licensing/src/class-manager.php

```diff
@@ -0,0 +1,195 @@
+<?php
+/**
+ * A Terms of Service class for Jetpack.
+ *
+ * @package automattic/jetpack-licensing
+ */
+
+namespace Automattic\Jetpack\Licensing;
+
+use Automattic\Jetpack\Connection\Manager as Connection_Manager;
+use Jetpack_IXR_ClientMulticall;
+use Jetpack_Options;
+use WP_Error;
+
+/**
+ * Class Manager.
+ *
+ * @since ??
+ *
+ * Helper class that is responsible for validating licenses to the current site.
+ */
+class Manager {
+	/**
+	 * Name of the WordPress option that holds all known Jetpack licenses.
+	 *
+	 * @const string
+	 */
+	const LICENSES_OPTION_NAME = 'jetpack_licenses';
+
+	/**
+	 * Holds the singleton instance of this class.
+	 *
+	 * @var self
+	 */
+	protected static $instance = false;
+
+	/**
+	 * Singleton.
+	 *
+	 * @static
+	 */
+	public static function instance() {
+		if ( ! self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Initialize.
+	 *
+	 * @return void
+	 */
+	public function initialize() {
+		add_action( 'update_option_' . self::LICENSES_OPTION_NAME, array( $this, 'validate_stored_licenses' ) );
+		add_action( 'jetpack_authorize_ending_authorized', array( $this, 'validate_stored_licenses_on_connection' ) );
+	}
+
+	/**
+	 * Get Jetpack connection manager instance.
+	 *
+	 * @return Connection_Manager
+	 */
+	protected function connection() {
+		return new Connection_Manager();
+	}
+
+	/**
+	 * Make an authenticated WP.com API multicall request instance.
+	 *
+	 * @param array $client_args IXR client arguments.
+	 * @return Jetpack_IXR_ClientMulticall
+	 */
+	protected function request( $client_args ) {
+		return new Jetpack_IXR_ClientMulticall( $client_args );
+	}
+
+	/**
+	 * Get all stored licenses.
+	 *
+	 * @return string[] License keys.
+	 */
+	public function stored_licenses() {
+		$licenses = (array) get_option( self::LICENSES_OPTION_NAME, array() );
+		$licenses = array_filter( $licenses, 'is_scalar' );
+		$licenses = array_map( 'strval', $licenses );
+		$licenses = array_filter( $licenses );
+
+		return $licenses;
+	}
+
+	/**
+	 * Validate the given licenses.
+	 *
+	 * @param string[] $licenses Licenses to validate.
+	 * @return array|WP_Error Results for each license (which may include WP_Error instances) or a WP_Error instance.
+	 */
+	public function validate_licenses( array $licenses ) {
+		if ( ! $this->connection()->is_active() ) {
+			return new WP_Error( 'not_connected', __( 'Jetpack is not connected.', 'jetpack' ) );
+		}
+
+		if ( empty( $licenses ) ) {
+			return array();
+		}
+
+		$xml = $this->request( array( 'user_id' => JETPACK_MASTER_USER ) );
+
+		foreach ( $licenses as $license ) {
+			$xml->addCall( 'jetpack.validateLicense', $license );
+		}
```

Comment:
Is this somewhere where we could consider using a v2 REST API endpoint instead of a new XML-RPC call?

You'll find examples in `_inc/lib/core-api/wpcom-endpoints/`; `_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-mailchimp.php` for instance fetches data from WordPress.com.

We happened to use a v2 endpoint as part of the connection flow up until very recently, so that could be another example to look at: 0d33951980a4cfcdf8a96b51073dcc4520564dde